### DISCUSSION TITLE
How to access the service interface to set its property?

### DISCUSSION THREAD

**Discussion by ruslan-mogilevskiy** [canton7/RestEase#240] [Q&A] [answered]
I have an API interface with a property that acts as an HTTP header value source:
`public interface IApi
    {
        [Header("X-Header")]
        string AuthHeader { get; set; }
    }`

and register it with the _services.AddRestEaseClient<IApi>_("base address").
The _AuthHeader_ value is known on the client registration time.

**Q**: when calling the AddRestEaseClient how to access the _IApi_ instance (which is generated with RestEase.For<IApi> under the hood) to set its AuthHeader property value?

**Comment by netclectic** [comment]
See the 2nd example under the Using HttpClientFactory section of the readme...

https://github.com/canton7/RestEase#using-httpclientfactory

**Comment by canton7** [comment]
Good question, there's no way currently. It's not too hard to add technically, but the overloads are becoming a bit of a mess and it's hard to add extra parameters in a way which doesn't break backwards compat.

**Reply by ruslan-mogilevskiy** [reply]
> Good question, there's no way currently. It's not too hard to add technically, but the overloads are becoming a bit of a mess and it's hard to add extra parameters in a way which doesn't break backwards compat.

It could be quite easily added and it will be backward compatible. I've switched to the UseWithRestEaseClient but it's also applicable to the AddRestEaseClient.

**HttpClientFactoryExtensions.cs**

public static IHttpClientBuilder UseWithRestEaseClient<T>(
            this IHttpClientBuilder httpClientBuilder,
            Action<RestClient>? configurer = null,
            **Action<T>? proxyConfigurer = null,**
            RequestModifier? requestModifier = null)
            where T : class
        {
            return httpClientBuilder.AddRestEaseClientCore(typeof(T), GenericFactory<T>(configurer, **proxyConfigurer**), requestModifier);
        }

        static Func<HttpClient, object> GenericFactory<T>(Action<RestClient> configurer, **Action<T> proxyConfigurer**) where T : class
        {
            return httpClient =>
            {
                var restClient = new RestClient(httpClient);
                configurer?.Invoke(restClient);
                var proxy = restClient.For<T>();
                **proxyConfigurer?.Invoke(proxy);**
                return proxy;
            };
        }

* sorry for the formatting, it's tricky to make proper code formatting on GitHub.

**Reply by canton7** [reply]
Right, but that change you made is a binary break -- people compiled against the old version of that method signature will now fail at runtime. I ended up taking this approach: https://github.com/canton7/RestEase/commit/f80748c1d85880d25f6d634daa66d8a113d0b8f8

**Reply by canton7** [reply]
I'll try and release that tomorrow

**Comment by ruslan-mogilevskiy** [comment]
Currently, I've fixed this problem by adding the header explicitly in the standard ConfigureHttpClient() but smashes the API interface setup across the application (API methods are in the _IApi_ interface while its auth headers are outside of it).

**Comment by canton7** [comment, marked as answer]
Released as 1.6.0
